src/Common/MSBuildCachePluginBase.cs

```diff
@@ -1145,6 +1093,48 @@ private void CheckForDuplicateOutputs(PluginLoggerBase logger, IReadOnlyDictiona
             logger.LogMessage($"Node {nodeContext.Id} produced duplicate-identical output {relativeFilePath} which was already produced by another node {previousNode.Id}. Allowing as content is the same.");
         }
     }
+
+    private static async Task<T> TimeAndLogAsync<T>(PluginLoggerBase? logger, Func<Task<T>> innerAsync, string? context = null, [CallerMemberName] string memberName = "")
+    {
+        var timer = Stopwatch.StartNew();
+        try
+        {
+            return await innerAsync();
+        }
+        catch (Exception e)
+        {
+            if (context == null)
+            {
+                logger?.LogWarning($"{memberName} failed after {timer.ElapsedMilliseconds} ms: {e}");
+            }
+            else
+            {
+                logger?.LogWarning($"{memberName}({context}) failed after {timer.ElapsedMilliseconds} ms: {e}");
+            }
+            throw;
+        }
+        finally
+        {
+            // only log 1+ second operations to avoid debug spew
+            if (timer.ElapsedMilliseconds > 1000.0)
+            {
+                if (context == null)
+                {
+                    logger?.LogMessage($"{memberName} succeeded after {timer.ElapsedMilliseconds} ms.");
+                }
+                else
+                {
+                    logger?.LogMessage($"{memberName}({context}) succeeded after {timer.ElapsedMilliseconds} ms.");
+                }
+            }
+        }
+    }
+
+    private static Task<int> TimeAndLogAsync(PluginLoggerBase? logger, Func<Task> innerAsync, string? context = null, [CallerMemberName] string memberName = "")
+        => TimeAndLogAsync(logger, async () => { await innerAsync(); return 0; }, context, memberName);
+
+    private static void TimeAndLog(PluginLoggerBase? logger, Action inner, string? context = null, [CallerMemberName] string memberName = "")
+        => TimeAndLogAsync(logger, () => { inner(); return Task.CompletedTask; }, context, memberName).SyncResult();
```

Comment:
Use `.GetAwaiter().GetResult()` instead of this extension. It's more clear and doesn't entangle more dependency on AzDO helpers.